feat/coverage

bogdan-manole opened the pull request:
## Description

Parse and add coverage to the `run-tests` result

Fixes #20 
Fixes #2  

Instructions:
- the output for `test-results` command has changed from an array of test results to an object embeding that list (`results` property) 
- to test the tool it should be used against `minimal-ptt-examples` on the branch `escrow-coverage` https://github.com/input-output-hk/minimal-ptt-examples/tree/escrow-coverage

bogdan-manole committed:
feat: add coverage support

KJES4 commented:
The `list-tests` command is returning this result that includes a test named coverage:
`[{"elements":[{"elements":[{"kind":"test","name":"can pay"},{"kind":"test","name":"can redeem"},{"kind":"test","name":"can redeem even if more money than required has been paid in"},{"kind":"test","name":"can refund"},{"kind":"test","name":"coverage"},{"kind":"test","name":"QuickCheck ContractModel"}],"kind":"group","name":"escrow"}],"kind":"group","name":"use cases"}]`

However the `run-tests` command does not return a result for a test named coverage. This is what is returned for the test results:
`,"results":[{"duration":8.0e-2,"name":"can pay","status":"OK"},{"duration":0.16,"name":"can redeem","status":"OK"},{"duration":0.19,"name":"can redeem even if more money than required has been paid in","status":"OK"},{"duration":0.1,"name":"can refund","status":"OK"},{"duration":16.15,"name":"QuickCheck ContractModel","status":"OK"}]}`

Is the test named coverage not supposed to show up in the list-tests returned output or is it missing from the run-tests output?

KJES4 commented:
1. There is an item called `coverage` showing up in the test tree when I list tests but no results are returned for this item when tests are run. I am not sure if this test item is supposed to be in the tree or not. 
2. Different results are being returned for the covered lines when I run the tests. Is this the intended response or is this unexpected behavior? Sometimes 8 lines are covered, sometimes, 12, sometimes 16, etc.
3. The covered and uncovered lines are overlapping. I don't believe this is the functionality we want.
4. The uncovered lines that returned are also repeating the same lines in this array multiple times.
5. When I run an individual test with the --test flag no items return in the covered array.

bogdan-manole commented:
> 1. There is an item called `coverage` showing up in the test tree when I list tests but no results are returned for this item when tests are run. I am not sure if this test item is supposed to be in the tree or not.
> 2. Different results are being returned for the covered lines when I run the tests. Is this the intended response or is this unexpected behavior? Sometimes 8 lines are covered, sometimes, 12, sometimes 16, etc.
> 3. The covered and uncovered lines are overlapping. I don't believe this is the functionality we want.
> 4. The uncovered lines that are returned are also repeating the same lines in this array multiple times.
> 5. When I run an individual test with the --test flag no items return in the covered array.

1. the `coverage` test is a leftover from some old attempts. I will remove it in the future, but for the moment, treat it as if it is a regular test
2. the difference comes from the fact that randomly generated values/paths are taken from the Quickcheck library when property testing is applied
3. for me as well the covered vs uncovered has some inconsistencies, I have to talk with other team members to understand why the QC-dynamic is representing the coverage report in such a way 
4. same as 3
5. can you please give me a real example?

KJES4 commented:
Nice! I figured that might be the case for point #2, I just wanted to confirm. 
For point 5: When I run this command `cabal run ptt-cli -- --project-path=../minimal-ptt-examples/ run-tests --test "use cases.escrow.can redeem"` for a single test there is no covered lines that are returned.

bogdan-manole committed:
feat: merge multiple coverage sections & add nix support

- multiple coverage code sections were presented with different status, now we merged them into one entry with the proper relevant status (coverage-status & ignore-status)
- add
🤟 [1;31mWelcome to nix-shell[0m 🤟

Type 'info' to see what's inside this shell. support (iogx)

bogdan-manole commented:
> covered

That's because the `redeem`  test doesn't support coverage momentarily

KJES4 reviewed (changes requested):
![image](https://github.com/user-attachments/assets/1d2c7c7b-101d-4cf8-b654-186049b874e9)

I am getting an error with the new changes and this is preventing the extension from loading the tests.

bogdan-manole committed:
test: coverage support tests

RSoulatIOHK reviewed (approved):
Approved but couldn't test it :( Need to reinstall nix and it's a nightmare...

bogdan-manole commented on src/Cardano/PTT/CLI/Process.hs:
```diff
@@ -138,10 +141,69 @@ readTestResults verbose h acc = liftIO $ do
+parseCoverageEntry str =
+  -- without status: src/Contract/Escrow.hs:490,22-490,55
+  -- with status: src/Contract/Escrow.hs:490,22-490,55 = False
+  -- let's parse with regex
```
I don't exactly remember if I used Copilot or ChatGPT (same engine, though) for the regex patterns, but for sure I used one of those 😄 .
The reason I left those comments there is that the reader needs to grasp at least what those patterns should match
